drivers/encx24j600/encx24j600.c

```diff
@@ -0,0 +1,424 @@
+
+    lock(dev);
+
+    /* initialization procedure as described in data sheet (39935c.pdf) */
+    do {
+        do {
+            vtimer_usleep(ENCX24J600_INIT_DELAY);
+            reg_set(dev, EUDAST, 0x1234);
+            vtimer_usleep(ENCX24J600_INIT_DELAY);
+        } while (reg_get(dev, EUDAST) != 0x1234);
+
+        while (!(reg_get(dev, ESTAT) & CLKRDY));
+
+        /* issue System Reset */
+        cmd(dev, SETETHRST);
+
+        /* make sure initialization finalizes */
+        vtimer_usleep(1000);
+    } while (!(reg_get(dev, EUDAST) == 0x0000));
+
+    /* configure flow control */
+    phy_reg_set(dev, PHANA, 0x05E1);
+    reg_set_bits(dev, ECON2, AUTOFC);
+
+    /* setup receive buffer */
+    reg_set(dev, ERXST, RX_BUFFER_START);
+    reg_set(dev, ERXTAIL, RX_BUFFER_END);
+    dev->rx_next_ptr = RX_BUFFER_START;
+
+    /* setup interrupts */
+    reg_set_bits(dev, EIE, PKTIE | LINKIE);
+    cmd(dev, ENABLERX);
+    cmd(dev, SETEIE);
+
+    DEBUG("encx24j600: initialization complete.\n");
+
+    unlock(dev);
+
+    return 0;
+}
+
+static int _send(dev_eth_t *encdev, char *pkt, int len) {
+    encx24j600_t * dev = (encx24j600_t *) encdev;
+    lock(dev);
+
+    /* wait until previous packet has been sent */
+    while ((reg_get(dev, ECON1) & TXRTS));
+
+    /* copy packet to SRAM */
+    sram_op(dev, WGPDATA, TX_BUFFER_START, pkt, len);
+
+    /* set start of TX packet and length */
+    reg_set(dev, ETXST, TX_BUFFER_START);
+    reg_set(dev, ETXLEN, len);
+
+    /* initiate sending */
+    cmd(dev, SETTXRTS);
+
+    /* wait for sending to complete */
+    //while ((reg_get(dev, ECON1) & TXRTS));
+
+    unlock(dev);
+
+    return len;
+}
+
+/*static void encx24j600_handle_linkstate(encx24j600_t *dev, int state)
+{
+    DEBUG("encx24j600: link %s.\n", state ? "UP" : "DOWN");
+}*/
+
+static inline int packets_available(encx24j600_t *dev) {
+    /* return PKTCNT (low byte of ESTAT) */
+    return reg_get(dev, ESTAT) & ~0xFF00;
+}
+
+static void _get_mac_addr(dev_eth_t *encdev, uint8_t* buf)
+{
+    encx24j600_t * dev = (encx24j600_t *) encdev;
+    uint16_t *addr = (uint16_t *) buf;
+
+    lock(dev);
+
+    addr[0] = reg_get(dev, MAADR1);
+    addr[1] = reg_get(dev, MAADR2);
+    addr[2] = reg_get(dev, MAADR3);
+
+    unlock(dev);
+}
+
+static int _recv(dev_eth_t *encdev, char* buf, int len)
+{
+    encx24j600_t * dev = (encx24j600_t *) encdev;
+    encx24j600_frame_hdr_t hdr;
+
+    lock(dev);
+
+    /* read frame header */
+    sram_op(dev, RRXDATA, dev->rx_next_ptr, (char*)&hdr, sizeof(hdr));
+
+    /* read packet */
+    sram_op(dev, RRXDATA, 0xFFFF, buf, hdr.frame_len);
+
+    /* decrement available packet count */
+    cmd(dev, SETPKTDEC);
+
+//    DEBUG("encx24j600: Received packet. frame_len=%u next=0x%x, newnext=0x%x ERXTAIL=0x%x\n", (unsigned int)hdr.frame_len, (unsigned int) dev->rx_next_ptr, (unsigned int)hdr.rx_next_ptr, (unsigned int)reg_get(dev, ERXTAIL));
```

Comment:
I found the debug message here is very helpful, I have changed a bit. https://github.com/jfischer-phytec-iot/RIOT/commit/b38a4c9c1519e938b23a7b3c48c721fa04567053